### PR TITLE
Add a parser for FASTA -m8CB, -m8CC output

### PR DESCRIPTION
**Pull request by mdehoon**
Add a parser to `Bio.Align` for output generated by the FASTA aligner with the `-m8CB` or `-m8CC` output format arguments. This is the currently recommended format for computer-parsable output.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/Align/fasta_m8.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
+# Copyright 2008-2016 by Peter Cock.  All rights reserved.
```
Comment (peterjc) [inline Bio/Align/fasta_m8.py:1]: This should be copyright you, but can acknowledge based on code by me. Or list both names as copyright holders as you prefer.

**Comment by peterjc** [comment]
So this is a sister format to the existing Pearson FASTA ``-m 10`` support in AlignIO,

https://github.com/biopython/biopython/blob/master/Bio/AlignIO/FastaIO.py
https://github.com/biopython/biopython/blob/master/Bio/SearchIO/FastaIO.py

I forget now how similar/different they are, but would sticking with ``-m 10`` not have worked well here?

And should these parsers cross reference each other in their docstrings?

**Comment by mdehoon** [comment]
@peterjc 
> So this is a sister format to the existing Pearson FASTA -m 10 support in AlignIO,
>
> https://github.com/biopython/biopython/blob/master/Bio/AlignIO/FastaIO.py
> 
> https://github.com/biopython/biopython/blob/master/Bio/SearchIO/FastaIO.py
> 
> I forget now how similar/different they are, but would sticking with -m 10 not have worked well here?

The documentation for the current version of FASTA says the following:

> -m 10 a parseable format for use with other programs (this option no longer reliably tested; -m
8CBL is easier to parse and tested more extensively).

so the -m 10 option is effectively deprecated.

File: Bio/Align/fasta_m8.py (hunk 1 above)
Comment (mdehoon) [reply Bio/Align/fasta_m8.py]: Replaced by my name, as the code is new.

File: Bio/Align/fasta_m8.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        Arguments:
+         - source   - input data or file name
+
+        """
+        super().__init__(source, mode="t", fmt="FASTA")
+        stream = self.stream
+        try:
+            line = next(stream)
+        except StopIteration:
+            raise ValueError("Empty file.") from None
```
Comment (peterjc) [inline Bio/Align/fasta_m8.py:49]: Not tested - I think I put in a blanket testing of all the SeqIO parsers with an empty file, as part of checking for consistent error messages. I don't know if we're at that point yet with the new alignment suite (no top level API)?

File: Bio/Align/fasta_m8.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        elif self._alignment_representation == "CIGAR":
+            coordinates = self.parse_cigar(columns[12])
+        coordinates[0, :] += target_start
+        if query_start < query_end:
+            coordinates[1, :] += query_start
+        else:
+            # mapped to reverse strand
+            coordinates[1, :] = query_start - coordinates[1, :] + 1
+        if self._query_size is None:
+            query_size = query_end
```
Comment (peterjc) [inline Bio/Align/fasta_m8.py:148]: This looks like an important corner case to have test coverage - do you have a test file which triggered it?

File: Bio/Align/fasta_m8.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+            length = int(length)
+            if operation == "M":
+                target_coordinate += length
+                query_coordinate += length
+            elif operation == "I":
+                target_coordinate += length
+            elif operation == "D":
+                query_coordinate += length
+            else:
+                raise ValueError("Unexpected operation '%s'" % operation)
```
Comment (peterjc) [inline Bio/Align/fasta_m8.py:234]: From the narrow regex above, is this possible?

Separately, M/D/I is only a tiny subset of the SAM/BAM CIGAR language. There is also =/X as an alternative to M, and S and H for clipping. Do you anticipate adding those as needed?

File: Bio/Align/fasta_m8.py (hunk 3 above)
Comment (mdehoon) [reply Bio/Align/fasta_m8.py:148]: With the current code, this cannot actually happen. I will remove it.

File: Bio/Align/fasta_m8.py (hunk 4 above)
Comment (mdehoon) [reply Bio/Align/fasta_m8.py:234]: To just define the alignment, M/D/I is sufficient, and I have not seen other operations in the cigar string in output generated by FASTA. We can add other operations if somebody finds them in FASTA output.

We may want to move this code to the `Alignment` class because it is generally useful, but I want to consider that at a later stage, once we have parsers for all alignment formats we want to support.

I'd like to keep the `ValueError` for now, to help with debugging if other operations indeed appear.

File: Bio/Align/fasta_m8.py (hunk 2 above)
Comment (mdehoon) [reply Bio/Align/fasta_m8.py:49]: I have added a test.

File: Bio/Align/fasta_m8.py (hunk 4 above)
Comment (mdehoon) [reply Bio/Align/fasta_m8.py]: Actually, the `ValueError` cannot appear. I will remove it.

Comment (peterjc) [reply Bio/Align/fasta_m8.py]: OK - if and when this is moved for use outside of the FASTA parser, then it would need to support other CIGAR operations.

**Review by peterjc — APPROVED** [review]

**Comment by mdehoon** [comment]
@peterjc  Thanks
